light-prover/prover/batch_update_circuit.go

```diff
@@ -0,0 +1,256 @@
+package prover
+
+import (
+	"fmt"
+	merkle_tree "light/light-prover/merkle-tree"
+	"light/light-prover/prover/poseidon"
+	"math/big"
+
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/constraint"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/frontend/cs/r1cs"
+	"github.com/reilabs/gnark-lean-extractor/v2/abstractor"
+)
+
+type BatchUpdateCircuit struct {
+	OldRoot             frontend.Variable `gnark:",public"`
+	NewRoot             frontend.Variable `gnark:",public"`
+	HashChainStartIndex frontend.Variable `gnark:",public"`
+	LeavesHashchainHash frontend.Variable `gnark:",public"`
+
+	Leaves       []frontend.Variable   `gnark:"input"`
+	MerkleProofs [][]frontend.Variable `gnark:"input"`
+	PathIndices  []frontend.Variable   `gnark:"input"`
+
+	Height    uint32
+	BatchSize uint32
+}
+
+func (circuit *BatchUpdateCircuit) Define(api frontend.API) error {
+	calculatedHashchainHash := circuit.createHashChain(api, int(circuit.BatchSize), circuit.Leaves)
+	api.AssertIsEqual(calculatedHashchainHash, circuit.LeavesHashchainHash)
+
+	api.AssertIsEqual(circuit.HashChainStartIndex, 0)
+
+	emptyLeaf := frontend.Variable(0)
+	newRoot := circuit.OldRoot
+
+	for i := 0; i < int(circuit.BatchSize); i++ {
+		indexBits := api.ToBinary(circuit.PathIndices[i], int(circuit.Height))
+		currentRoot := circuit.merkleRoot(api, circuit.Leaves[i], indexBits, circuit.MerkleProofs[i])
+		api.AssertIsEqual(currentRoot, newRoot)
+		newRoot = circuit.merkleRoot(api, emptyLeaf, indexBits, circuit.MerkleProofs[i])
+	}
+
+	api.AssertIsEqual(newRoot, circuit.NewRoot)
+
+	return nil
+}
+
+func (circuit *BatchUpdateCircuit) merkleRoot(api frontend.API, leaf frontend.Variable, indexBits []frontend.Variable, siblings []frontend.Variable) frontend.Variable {
```

Comment:
can we unify this with the verify Merkle proof implementation in circuit_utils?

Reply:
`MerkleRootGadget` in `circuit_utils` is equivalent with this gadget right?

Reply:
practically yes, I'll unify it